vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,513 @@
+)
+
+func osutilCommand(bin string, args ...string) subProcessCmd {
+	return osutil.Command(bin, args...)
+}
+
+type subProcessCmd interface {
+	StdinPipe() (io.WriteCloser, error)
+	StdoutPipe() (io.ReadCloser, error)
+	StderrPipe() (io.ReadCloser, error)
+	Start() error
+	Wait() error
+}
+
+func makeDefaultParams() *proxyAppParams {
+	return &proxyAppParams{
+		CommandRunner:   osutilCommand,
+		MakePipe:        osutil.LongPipe,
+		InitRetryDelay:  10 * time.Second,
+		MaxInitAttempts: math.MaxInt64,
+	}
+}
+
+func init() {
+	vmimpl.Register(
+		"proxyapp",
+		func(env *vmimpl.Env) (vmimpl.Pool, error) {
+			return ctor(makeDefaultParams(), env)
+		},
+		false)
+}
+
+type pool struct {
+	wgInit          sync.WaitGroup
+	isValid         bool
+	env             *vmimpl.Env
+	proxy           *ProxyApp
+	ctxRunner       context.Context
+	ctxRunnerCancel context.CancelFunc
+	size            int
+}
+
+func (pool *pool) reset() {
+	pool.cancelRun()
+	pool.ctxRunner, pool.ctxRunnerCancel = context.WithCancel(context.TODO())
+	pool.proxy = nil
+	pool.size = 0
+	pool.isValid = false
+}
+
+func (pool *pool) cancelRun() {
+	if pool.ctxRunnerCancel != nil {
+		pool.ctxRunnerCancel()
+		pool.ctxRunnerCancel = nil
+	}
+}
+
+func initPool(params *proxyAppParams, pool *pool, command string, pluginConfig []byte, env *vmimpl.Env,
+) error {
+	pool.reset()
+
+	var err error
+	pool.proxy, err = runProxyApp(pool.ctxRunner, params, command)
+	if err != nil {
+		return fmt.Errorf("failed to run ProxyApp: %w", err)
+	}
+
+	pool.size, err = pool.proxy.ConstructPool(string(pluginConfig), env.Debug)
+	if err != nil {
+		pool.cancelRun()
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	pool.isValid = true
+	return nil
+}
+
+type Config struct {
+	Command        string          `json:"proxy_cmd"`
+	ProxyAppConfig json.RawMessage `json:"proxyapp_config"`
+}
+
+func parseConfig(conf []byte) (*Config, error) {
+	vmCfg := new(Config)
+	if err := config.LoadData(conf, vmCfg); err != nil {
+		return nil, fmt.Errorf("failed to parseConfig(): %w", err)
+	}
+	return vmCfg, nil
+}
+
+// Package configuration VARs are mostly needed for tests.
+type proxyAppParams struct {
+	CommandRunner   func(bin string, args ...string) subProcessCmd
+	MakePipe        func() (io.ReadCloser, io.WriteCloser, error)
+	InitRetryDelay  time.Duration
+	MaxInitAttempts int64
+}
+
+func ctor(params *proxyAppParams, env *vmimpl.Env) (vmimpl.Pool, error) {
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	pool := &pool{
+		env: env,
+	}
+	pool.wgInit.Add(1)
+	minWait := time.After(0)
+	initialized := make(chan bool)
+	attempt := int64(0)
+
+	go func() {
+		for {
+			attempt++
+			if attempt > params.MaxInitAttempts {
+				pool.wgInit.Done()
+				select {
+				case initialized <- false:
+				default:
+				}
+
+				return
+			}
+
+			<-minWait
+			minWait = time.After(params.InitRetryDelay)
```

Comment:
There are two ends of the loop - "continue" and final end. Because of them minWait for at the start.
Adding "else" to forward both execution paths down and move <-minWait to the end.
I don't think it makes code more readable now. What do you think?

I like <-minWait more than Sleep() because it would make every at loop max(minWait, initPool_time).
Sleep(minWait) would make every loop (minWait + initPool_time).

Reply:
> With the current code we still can have busy loop if initPool_time >= minWait.
Example: minWait is 10 seconds. initPool_time is 10 minutes.
<-minWait will return immediately, Loop cycle is 10 minutes. No busy loop.

minWait works only if initPool_time is less that 10 seconds.

My approach looks confusing. Lets mark it as a premature optimization and use time.Sleep() instread.
Shorter, simpler.